### DISCUSSION TITLE
CT-e

### DISCUSSION THREAD

**Discussion by ibemoreira** [TadaSoftware/PyNFe#195] [General]
Alguém já utilizou essa biblioteca para baixar CT-e?

**Comment by ibemoreira** [comment]
Estava vendo que o cteDistribuicaoDFe é uma api separada dá NFeDistribuicaoDFe, essa lib já tem essa função?

**Comment by ibemoreira** [comment]
Siga tentando fazer a distribuição de CT-e debugando a lib e tentando consumir o  'https://www1.cte.fazenda.gov.br/CTeDistribuicaoDFe/CTeDistribuicaoDFe.asmx'

Sem sucesso até agora. 
Se poder dar uma dica os caminho das pedra @leogregianin  fico agradecido. 

Os xmls de NFe esta rodando a 6 dias no servidor da empresa sem problemas consumindo o serviço a cada 12 horas. 

Desde já muito obrigado estou muito feliz com resultado até agora (não sou programador).

**Comment by leogregianin** [comment]
@ibemoreira essa lib não tem nada relacionado ao CTe, ainda.

O serviço de Distribuição de CT-e é muito parecido com a Distribuição de NF-e.

Sugiro criar o dicionário de URLs do CTe [aqui](https://github.com/TadaSoftware/PyNFe/blob/main/pynfe/utils/webservices.py) conforme as [URLs do CT-e](https://www.cte.fazenda.gov.br/portal/webServices.aspx?tipoConteudo=wpdBtfbTMrw=).

Criar o método de Distribuição DFe para o CTe igual a NFe:
https://github.com/TadaSoftware/PyNFe/blob/2f12e171ccb35ff38dd8c106207704e7c8cb131b/pynfe/processamento/comunicacao.py#L166

Ler e entender bem as Notas Técnicas do CTe para esse tipo de serviço.
Nesta issue #88 tem links das notas técnicas, manuais e schemas.

**Comment by leogregianin** [comment]
A consulta da Distribuição de CTe foi feita e tem exemplo de uso nesse wiki: https://github.com/TadaSoftware/PyNFe/wiki/Fluxo-da-Consulta-de-Distribuicao-CTe

**Reply by ibemoreira** [reply]
@leogregianin  Bom dia, brother por um tempo o fluxo de Distribuição CTe funcionou, mas agora só devolve esse erro: 
`Traceback (most recent call last):
  File "c:\Projeto\PROTHEUS_SQL\Wilbert\CTe\baixar_cte_novo.py", line 48, in <module>
    resposta_descompactado = DescompactaGzip.descompacta(zip_resposta)
  File "C:\Users\Lucas\AppData\Local\Programs\Python\Python310\lib\site-packages\pynfe\utils\descompactar.py", line 30, in descompacta
    return etree.fromstring(descompactado)
  File "src\lxml\etree.pyx", line 3254, in lxml.etree.fromstring
  File "src\lxml\parser.pxi", line 1908, in lxml.etree._parseMemoryDocument
ValueError: Unicode strings with encoding declaration are not supported. Please use bytes input or XML fragments without declaration.`

**Reply by herbertsmatheus** [reply]
Boa tarde,
Estava tendo o mesmo problema que voce @ibemoreira, para resolver, editei a linha 30 do código do arquivo 'descompatar.py' de:
`return etree.fromstring(descompactado)`
para
`return etree.fromstring(bytes(descompactado, 'utf-8'))`

Está funcionando pra mim, porém não sei se essa mudança pode gerar problemas em outras ocasiões.

**Comment by ibemoreira** [comment]
Boa tarde, brother eu já até estou rodando ela aqui na empresa kkk, sou
iniciante  estou estudando python mas pynfe já me salvou por aqui, muitoo
obrigadooo!

Em ter., 9 de ago. de 2022 às 12:53, Leonardo Gregianin <
***@***.***> escreveu:

**Comment by leogregianin** [comment]
@ibemoreira, para melhor organização das discussões do projeto, pode marcar essa discussão por respondida?
